README.md
This is a fork of https://gitlab.com/alange/open_library

# open_library

A Dart implementation of openlibrary.org api.

## Features

* fetch book information using the https://openlibrary.org api ISBN
* supports 10 and 13 digit ISBN numbers
* adds authors and covers to the book result object
* covers can be easily used in flutter providing MemoryImages using Uint8List as covers
* search for books using the query methods for title, authors, isbn and unified queries
* new search patterns for subject, place, person, language, and publisher
* load covers in small, medium, large or even disable cover load

## Example App

<img src="https://gitlab.com/alange/open_library/-/raw/main/screenshot.png?inline=false" height="300">

## Install

In the `pubspec.yaml` of your flutter project, add the following dependency:

```yaml
dependencies:
  open_library: <latest_version>
```

In your library add the following import:

```dart
import 'package:open_library/open_library.dart';
```

## Getting started

 Example:
```dart
import 'package:flutter/material.dart';
import 'package:open_library/open_library.dart';
import 'package:provider/provider.dart';
import 'screens/book_screen.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Provider(
        create: (_) => OpenLibrary(),
        dispose: (_, OpenLibrary service) => service.dispose(),
        child: MaterialApp(
          title: 'Book Example',
          home: BookScreen()
        ));
  }
}

import 'package:flutter/material.dart';
import 'package:open_library/models/ol_book_model.dart';
import 'package:open_library/models/ol_search_model.dart';
import 'package:open_library/open_library.dart';
import 'package:provider/provider.dart';

const ISBN1 = '9783608980806';
const ISBN2 = '0674995171';
const ISBN3 = '3596191130';

class BookScreen extends StatefulWidget {
  BookScreen({Key? key}) : super(key: key);

  final List<OLBook> books = [];

  @override
  State<BookScreen> createState() => _BookScreenState();
}

class _BookScreenState extends State<BookScreen> {
  final TextEditingController controller =
      TextEditingController(text: "Lord of the Rings");
  late bool isLoading = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.blueGrey,
      body: !isLoading
          ? Column(
              children: [
                const SizedBox(height: 50.0),
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    SizedBox(
                      width: MediaQuery.of(context).size.width * 0.6,
                      height: 60.0,
                      child: TextField(
                        controller: controller,
                        cursorColor: Colors.white,
                        style: const TextStyle(color: Colors.white),
                        decoration: const InputDecoration(
                          label: Text("Book title:"),
                          labelStyle: TextStyle(color: Colors.white),
                          focusColor: Colors.white,
                        ),
                      ),
                    ),
                    Padding(
                      padding: const EdgeInsets.only(left: 10.0),
                      child: GestureDetector(
                        onTap: () async {
                          setState(() {
                            isLoading = true;
                          });
                          final OLSearchBase search =
                              await Provider.of<OpenLibrary>(context,
                                      listen: false)
                                  .query(q: controller.text, limit: 4);
                          if (search is OLSearch) {
                              widget.books.clear();
                              print("search:\n$search");
                              for (var doc in search.docs) {
                                final OLBook book = OLBook(
                                  title: doc.title,
                                  authors: doc.authors,
                                  covers: doc.covers,
                                );
                                widget.books.add(book);
                              }
                          }
                          setState(() {
                            isLoading = false;
                          });
                        },
                        child: const Icon(
                          Icons.search,
                          color: Colors.white,
                        ),
                      ),
                    ),
                  ],
                ),
                const SizedBox(
                  height: 20.0,
                ),
                const Center(
                    child: Text(
                      "press action button to search for those ISBN's",
                      style: TextStyle(color: Colors.white),
                    )),
                const SizedBox(
                  height: 20.0,
                ),
                const Center(
                    child: Text(
                  "ISBN1:$ISBN1",
                  style: TextStyle(color: Colors.white),
                )),
                const SizedBox(height: 20.0),
                const Center(
                    child: Text(
                  "ISBN2:$ISBN2",
                  style: TextStyle(color: Colors.white),
                )),
                const SizedBox(height: 20.0),
                const Center(
                    child: Text(
                  "ISBN3:$ISBN3",
                  style: TextStyle(color: Colors.white),
                )),
                const SizedBox(height: 20.0),
                SingleChildScrollView(
                  child: ListView.builder(
                    shrinkWrap: true,
                    itemCount: widget.books.length,
                    itemBuilder: (context, index) {
                      return bookWidget(
                          book: widget.books[index], context: context);
                    },
                  ),
                )
              ],
            )
          : const Center(child: CircularProgressIndicator(color: Colors.white,)),
      floatingActionButton: FloatingActionButton(
        onPressed: () async {
          setState(() {
            isLoading = true;
          });
          widget.books.clear();
          const bool loadCovers = true;
          const CoverSize size = CoverSize.S;
          final OLBookBase book1 =
              await Provider.of<OpenLibrary>(context, listen: false)
                  .getBookByISBN(
                      isbn: ISBN1, loadCover: loadCovers, coverSize: size);
          print(book1.toString());
          if (book1 is OLBook) {
            widget.books.add(book1);
          }
          final OLBookBase book2 =
              await Provider.of<OpenLibrary>(context, listen: false)
                  .getBookByISBN(
                      isbn: ISBN2, loadCover: loadCovers, coverSize: size);
          print(book2.toString());
          if (book2 is OLBook) {
            widget.books.add(book2);
          }
          final OLBookBase book3 =
              await Provider.of<OpenLibrary>(context, listen: false)
                  .getBookByISBN(
                      isbn: ISBN3, loadCover: loadCovers, coverSize: size);
          print(book3.toString());
          if (book3 is OLBook) {
            widget.books.add(book3);
          }
          setState(() {
            isLoading = false;
          });
        },
        child: const Icon(Icons.book),
      ),
    );
  }

  Widget bookWidget({required OLBook book, required BuildContext context}) {
    String author = '';
    if (book.authors.isNotEmpty) {
      author = book.authors.first.name.trim();
    }
    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: Container(
        width: MediaQuery.of(context).size.width * 0.8,
        height: 80.0,
        decoration: BoxDecoration(
          color: Colors.white,
          borderRadius: BorderRadius.circular(10.0),
        ),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Column(
              mainAxisAlignment: MainAxisAlignment.center,
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Padding(
                  padding:
                      const EdgeInsets.only(top: 4.0, bottom: 4.0, left: 20.0),
                  child: SizedBox(
                    width: MediaQuery.of(context).size.width * 0.6,
                    child: Text(
                      book.title,
                      overflow: TextOverflow.ellipsis,
                      style: const TextStyle(
                          color: Colors.black,
                          fontSize: 14,
                          fontWeight: FontWeight.bold),
                    ),
                  ),
                ),
                Padding(
                  padding:
                      const EdgeInsets.only(top: 4.0, bottom: 4.0, left: 20.0),
                  child: Text(
                    author,
                    style: const TextStyle(color: Colors.black, fontSize: 12),
                  ),
                ),
              ],
            ),
            Padding(
              padding: const EdgeInsets.only(right: 20.0),
              child: SizedBox(
                height: book.covers.isNotEmpty ? 64.0 : 0,
                child: book.covers.isNotEmpty
                    ? Image.memory(book.covers.first)
                    : null,
              ),
            ),
          ],
        ),
      ),
    );
  }
}

```



CHANGELOG.md
## 0.1.6
* Remove Flutter dependancies. Now package is pure dart

## 0.1.5
* Bugfix: Error in pattern builder to render search patterns without title now works

## 0.1.4
* updated doc to follow the dart guidelines to create more readable docs described at https://dart.dev/guides/language/effective-dart/documentation

## 0.1.3
* updated project structure to separate private code from public api code

## 0.1.2
* new search patterns for subject, place, person, language, and publisher
* you now can fully combine search patterns

## 0.1.1
* documentation update

## 0.1.0
* New features to support the search.json query engine of openlibrary.org
* New Models to support rich property models for search result
* selecting L/M/S covers on search to reduce loading times for previews
* getBookByISBN now supports cover loading true/false
* getBookByISBN now supports cover size S/M/L on fetch
* Covers can now be loaded independently
This version is the first bump to more features on the openlibrary.org search

## 0.0.3
* Documentation Update
documentation embedded image fixed.

## 0.0.2
* Documentation Update
some updates on the documentation.

## 0.0.1

* Initial Release.
the initial release only supports to fetch books via it's ISBN (10,13 digits). It provides
  a bunch of models to represent books and authors. The OLBook Model will give you access
  to almost all data provided by openlibrary.org isbn/book api. Also it adds the authors and
  covers as Uint8List if covers exists.
  


example/lib/screens/book_screen.dart
import 'package:flutter/material.dart';
import 'package:open_library/models/ol_book_model.dart';
import 'package:open_library/models/ol_search_model.dart';
import 'package:open_library/open_library.dart';
import 'package:provider/provider.dart';

const ISBN1 = '9783608980806';
const ISBN2 = '0674995171';
const ISBN3 = '3596191130';

class BookScreen extends StatefulWidget {
  BookScreen({Key? key}) : super(key: key);

  final List<OLBook> books = [];

  @override
  State<BookScreen> createState() => _BookScreenState();
}

class _BookScreenState extends State<BookScreen> {
  final TextEditingController controller =
      TextEditingController(text: "Lord of the Rings");
  late bool isLoading = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.blueGrey,
      body: !isLoading
          ? Column(
              children: [
                const SizedBox(height: 50.0),
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    SizedBox(
                      width: MediaQuery.of(context).size.width * 0.6,
                      height: 60.0,
                      child: TextField(
                        controller: controller,
                        cursorColor: Colors.white,
                        style: const TextStyle(color: Colors.white),
                        decoration: const InputDecoration(
                          label: Text("Book title:"),
                          labelStyle: TextStyle(color: Colors.white),
                          focusColor: Colors.white,
                        ),
                      ),
                    ),
                    Padding(
                      padding: const EdgeInsets.only(left: 10.0),
                      child: GestureDetector(
                        onTap: () async {
                          setState(() {
                            isLoading = true;
                          });
                          final OLSearchBase search =
                              await Provider.of<OpenLibrary>(context,
                                      listen: false)
                                  .query(title: controller.text, limit: 4);
                          if (search is OLSearch) {
                              widget.books.clear();
                              print("search:\n$search");
                              for (var doc in search.docs) {
                                final OLBook book = OLBook(
                                  title: doc.title,
                                  authors: doc.authors,
                                  covers: doc.covers,
                                );
                                widget.books.add(book);
                              }
                          }
                          setState(() {
                            isLoading = false;
                          });
                        },
                        child: const Icon(
                          Icons.search,
                          color: Colors.white,
                        ),
                      ),
                    ),
                  ],
                ),
                const SizedBox(
                  height: 20.0,
                ),
                const Center(
                    child: Text(
                      "press action button to search for those ISBN's",
                      style: TextStyle(color: Colors.white),
                    )),
                const SizedBox(
                  height: 20.0,
                ),
                const Center(
                    child: Text(
                  "ISBN1:$ISBN1",
                  style: TextStyle(color: Colors.white),
                )),
                const SizedBox(height: 20.0),
                const Center(
                    child: Text(
                  "ISBN2:$ISBN2",
                  style: TextStyle(color: Colors.white),
                )),
                const SizedBox(height: 20.0),
                const Center(
                    child: Text(
                  "ISBN3:$ISBN3",
                  style: TextStyle(color: Colors.white),
                )),
                const SizedBox(height: 20.0),
                SingleChildScrollView(
                  child: ListView.builder(
                    shrinkWrap: true,
                    itemCount: widget.books.length,
                    itemBuilder: (context, index) {
                      return bookWidget(
                          book: widget.books[index], context: context);
                    },
                  ),
                )
              ],
            )
          : const Center(child: CircularProgressIndicator(color: Colors.white,)),
      floatingActionButton: FloatingActionButton(
        onPressed: () async {
          setState(() {
            isLoading = true;
          });
          widget.books.clear();
          const bool loadCovers = true;
          const CoverSize size = CoverSize.S;
          final OLBookBase book1 =
              await Provider.of<OpenLibrary>(context, listen: false)
                  .getBookByISBN(
                      isbn: ISBN1, loadCover: loadCovers, coverSize: size);
          print(book1.toString());
          if (book1 is OLBook) {
            widget.books.add(book1);
          }
          final OLBookBase book2 =
              await Provider.of<OpenLibrary>(context, listen: false)
                  .getBookByISBN(
                      isbn: ISBN2, loadCover: loadCovers, coverSize: size);
          print(book2.toString());
          if (book2 is OLBook) {
            widget.books.add(book2);
          }
          final OLBookBase book3 =
              await Provider.of<OpenLibrary>(context, listen: false)
                  .getBookByISBN(
                      isbn: ISBN3, loadCover: loadCovers, coverSize: size);
          print(book3.toString());
          if (book3 is OLBook) {
            widget.books.add(book3);
          }
          setState(() {
            isLoading = false;
          });
        },
        child: const Icon(Icons.book),
      ),
    );
  }

  Widget bookWidget({required OLBook book, required BuildContext context}) {
    String author = '';
    if (book.authors.isNotEmpty) {
      author = book.authors.first.name.trim();
    }
    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: Container(
        width: MediaQuery.of(context).size.width * 0.8,
        height: 80.0,
        decoration: BoxDecoration(
          color: Colors.white,
          borderRadius: BorderRadius.circular(10.0),
        ),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Column(
              mainAxisAlignment: MainAxisAlignment.center,
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Padding(
                  padding:
                      const EdgeInsets.only(top: 4.0, bottom: 4.0, left: 20.0),
                  child: SizedBox(
                    width: MediaQuery.of(context).size.width * 0.6,
                    child: Text(
                      book.title,
                      overflow: TextOverflow.ellipsis,
                      style: const TextStyle(
                          color: Colors.black,
                          fontSize: 14,
                          fontWeight: FontWeight.bold),
                    ),
                  ),
                ),
                Padding(
                  padding:
                      const EdgeInsets.only(top: 4.0, bottom: 4.0, left: 20.0),
                  child: Text(
                    author,
                    style: const TextStyle(color: Colors.black, fontSize: 12),
                  ),
                ),
              ],
            ),
            Padding(
              padding: const EdgeInsets.only(right: 20.0),
              child: SizedBox(
                height: book.covers.isNotEmpty ? 64.0 : 0,
                child: book.covers.isNotEmpty
                    ? Image.memory(book.covers.first)
                    : null,
              ),
            ),
          ],
        ),
      ),
    );
  }
}
